### ISSUE TITLE
Fix the Nesterov bug

### ISSUE THREAD

**Issue by oysteijo** [oysteijo/simd_neuralnet#17]
The nesterov code has a bad bug. It does an intermediate parameter update and then calculate the gradient based on these new weights. However, the real update is also adding the same parameter updates!??

There are several solutions to this.

**Comment by oysteijo** [comment]
Maybe this is simpler then I think.... If I read the code right is is simply just the order of the operations that is different

Without nesterov (but with momentum) we do:

- Calculate the gradient.
- Update the velocity: v_new = momentum * v - lr * g
- Update network weights: w_new = w + v_new

With Nesterov we do:
- Update (interim) network weights:  w_new = w + momentum * v
- Calculate the gradient (g) based on these new weights.
- Update the velocity: v_new = momentum * v - lr * g
- Update network weights: w_new = w + v_new

however w_new is then expressed: w_new = w + momentum * v - lr * g, and momentum * v has already been applied in the interim update. That means, I can probably do this:

- Update (interim) network weights:  w_new = w + momentum * v
- Calculate the gradient (g) based on these new weights.
- Update the velocity: v_new = momentum * v - lr * g
- Update network weights: w_new = w - lr * g

(I think so.... however I have to think to make this general. Maybe I should split this code into several files?)

**Comment by oysteijo** [comment]
Yes! It is probably fu*king easy. I just need to keep track of two parameter arrays.

- momentum * velocity
- -learning_rate * g

Wish me good luck!

**Comment by oysteijo** [comment]
24a2fd2 seems to fix the issues, but make a good test case and compare to a known implementation ie. Keras.
